### DISCUSSION TITLE
--username option usage

### DISCUSSION THREAD

**Discussion by son9701** [esnet/iperf#1455] [Q&A] [answered]
I want to use --useroption but I don't know how to use it.
Do I need to set it up on the server before the client can use it?

![image](https://user-images.githubusercontent.com/66204516/212288413-4c857436-b86d-47f7-a088-1822d07fc2cd.png)

and Is there a reason why I can't see the --username option with -h in a Linux environment?
![image](https://user-images.githubusercontent.com/66204516/212289344-4cf3cc81-33be-4b88-859c-e5a896c6bd30.png)

**Comment by davidBar-On** [comment]
`--username` is available only if an OpenSSL library was detected during the iperf3 build.  If you built iperf3 yourself search for "ssl" in `./config.log` (created when your run `./configure`) for a related error/warning messages.

If OpenSSL library library is available, you may try building iperf3 using `./configure --with-openssl=<root of the OpenSSL directory>`.

**Reply by son9701** [reply]
Now build with openssl and you will see that option
I have a few more questions
1. --rsa-public-key-path public.pem --username @@ It starts when any value is entered. Is there a way to set the username and password?
2. Can I use --username only with --rsa-public-key-path public.pem ?
3. The option is a client option, but the server cannot set username and password?
Kindly explain please

**Comment by davidBar-On** [comment, marked as answer]
On the **client** side, the user name is set by the`--username` and the password is set either by the `IPERF3_PASSWORD` environment variable, or iperf3 will prompt for it.  When using `--username`, `--rsa-public-key-path` **must** be set also.

On the **server** side, the following two parameters must be set for the authorization to work:
1. `--rsa-private-key-path` - the private key path.
2. `--authorized-users-path` - the path to the file that includes the user names and their password.  [This Oracle page](https://docs.oracle.com/cd/E88353_01/html/E37839/iperf3-1.html) has a good explanation about this file format and an example of generating it (see both the options descriptions and the EXAMPLES section).

**Reply by son9701** [reply]
Thanks for the kind explanation, and i have one more question.

I tried to create a key on the server and test it as explained on the Oracle page, but at least any path after --authorized-users-path is executed. is this normal?

And when I try to execute it with the values ​​set by S_ USER and S_ PASSWD, iperf cannot be executed with authorization fail. How can I solve this?
![image](https://user-images.githubusercontent.com/66204516/215363750-2809212c-6308-4eb1-9057-8a8eb510042b.png)

**Comment by davidBar-On** [comment]
I assume the "a" file is the "credentials.csv" file in the Oracle example.  Is this correct?  (Note that they have a typo there: the "cat credentials.csv" should be "cat >credentials.csv".)

Did you create the corresponding "public.pem" file and set the `--username` and `--rsa-public-key-path` when you run the client?  Note that the username and its password should be the same as you defined for the "a" file.

**Reply by son9701** [reply]
The file named "a" does not exist, but it worked and left a question.
![image](https://user-images.githubusercontent.com/66204516/215438833-1e3db0db-2695-4d1f-8349-e6874a44079c.png)

and when > S_USER=mario S_PASSWD=rossi was typed, file named S_USER=mario is created. Can I modify and use this file?

plus PEM file was created in order openssl genrsa -des3 -out private.pem 2048, openssl rsa -in private.pem -outform PEM -pubout -out public.pem

**Comment by davidBar-On** [comment]
Seems that you have not created and used the authorized users file correctly.  You should first create the file.  In the Oracle example the username is "mario" and the password is "rossi".  Per that example, first create the password hash:
```
> S_USER=mario S_PASSWD=rossi
> echo -n "{$S_USER}$S_PASSWD" | sha256sum | awk '{ print $1 }'
```
Then create the authorized users file "credentials.csv", where "bf7a49a846d44b454a5d11e7acfaf13d138bbe0b7483aa3e050879700572709b" is the password hash from the above:
```
> cat >credentials.csv
# file format: username,sha256
mario,bf7a49a846d44b454a5d11e7acfaf13d138bbe0b7483aa3e050879700572709b
```
Now, run the server using `--authorized-users-path credentials.csv` and the client using `--username mario".  Answer "rossi" the client's prompt for password.

**Reply by son9701** [reply]
Thanks, it's solved.
